### PR TITLE
Service and dockerization overhaul

### PR DESCRIPTION
**Pull request by edouard-sn**
Better python logging, working recommendation service and better multi-staged Dockerfile.
Added shebangs to scripts.
Requirements.txt cleaned up

Don't know if there will still be a `codegen` makefile rule, so I kept `misc/Dockerfile`

still TODO:
- Separate into different python files
- Docs on env variables

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
+
+
+# placeholder
+service-run:
+	docker run noted-recommendations-service -d -p 8080:8080
+# placeholder
+
+
+
+
```
Comment (rojas-diego) [inline Makefile:21]: Same thing, you should configure your editor to trim excessive newlines at the end of a file.

**Review by rojas-diego — CHANGES_REQUESTED** [review]
Requested some changes.

*Note:* Could you include a simple doc in the readme or elsewhere to indicate how to run the service outside of Docker? What's required etc.

File: recommendations-service.py [hunk 2]
```diff
@@ -1,3 +1,87 @@
...
+
+        logger.info("Recommendation service has been successfully initialized")
+
+    def __candidate_selection_and_weighting(self):
+        logger.debug("Candidate selection and weighting being processed")
+        self.extractor.candidate_selection(n=self.n_gram_length)
+        self.extractor.candidate_weighting(window=self.co_occurence_window, use_stems=False)
+
+    def ExtractKeywords(self, request, context):
+        logger.info("ExtractKeywords request being processed")
```
[the remark this answers is not in the document]
Comment (edouard-sn) [reply recommendations-service.py:33]: Noted as a TODO for now

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# gRPC deps
+grpcio-tools
+googleapis-common-protos
+
+python-dotenv # dotenv
+
+git+https://github.com/boudinfl/pke.git # Keyword Extraction Toolkit
+
+loguru
```
[the remark this answers is not in the document]
Comment (edouard-sn) [reply requirements.txt]: `pip freeze` dumps all the current env's python packages with versions, `pipreqs` tries to do the same with a project-specific scope but doesn't work with some usecases (from what I saw, ours as well).
New commits specifies packages versions

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+python-dotenv==0.20.0
+loguru==0.6.0
+
+git+https://github.com/boudinfl/pke.git # Keyword Extraction Toolkit
+
+# gRPC deps
+grpcio-tools==1.46.3
+googleapis-common-protos==1.56.1
```
Comment (rojas-diego) [inline requirements.txt:8]: Regarding what you said about how `pip freeze` dumps all the packages in your env, I believe if you use a virtual environment it dumps everything project-scoped nah? I may be wrong, I don't know much about Python.

Comment (rojas-diego) [reply requirements.txt:8]: Also, thank you for adding the versions.

**Review by rojas-diego — CHANGES_REQUESTED** [review]

**Comment by rojas-diego** [comment]
Also could be cool if you added the RPC logging middleware.

File: requirements.txt (hunk 4 above)
Comment (edouard-sn) [reply requirements.txt:8]: You could, but everything should be installed before hand you call `pip freeze`, and we already use this requirements.txt to install everything before hand.

Comment (rojas-diego) [reply requirements.txt:8]: I don't think we're talking about the same thing. Let's go over this next time.

Comment (edouard-sn) [reply requirements.txt:8]: I mean, i guess you're talking about python venv, but even if we use it we need to specify which packages to install, here, I chose to do so with a requirements.txt.

**Review by rojas-diego — APPROVED** [review]

**Pull request merged**